File: 46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#

# @lc code=start
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def backtrack(tmp, visited):
            if len(tmp) == len(nums):
                res.append(tmp.copy())
                return
            for i in range(len(nums)):
                if visited[i]:
                    continue
                tmp.append(nums[i])
                visited[i] = True
                backtrack(tmp, visited)
                visited[i] = False
                tmp.pop()

        backtrack([], [False]*len(nums))
        return res
            

# @lc code=end

File: 46.全排列.java
/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        order(nums, new ArrayList<>(), visited, res);
        return res;
    }
    public void order(int[] nums, List<Integer> tmp, boolean[] visited, List<List<Integer>> res){
        if(tmp.size()==nums.length){
            res.add(new ArrayList<>(tmp));
            return;
        }
        for(int i=0;i<nums.length;i++){
            if(visited[i])continue;
            visited[i] = true;
            tmp.add(nums[i]);
            order(nums, tmp, visited, res);
            tmp.remove(tmp.size()-1);
            visited[i] = false;
        }
    }

}
// @lc code=end

